### PR TITLE
Fix cross series aggregates logic

### PR DESCRIPTION
**Pull request by talIguaz**

### CODE REVIEW

File: pkg/pquerier/collector.go [hunk 1]
```diff
@@ -72,16 +74,25 @@ func mainCollector(ctx *selectQueryContext, responseChannel chan *qryResults) {
...
-			} else if res.IsClientAggregates() {
-				aggregateClientAggregates(ctx, res)
+			lsetAttr, _ := res.fields[config.LabelSetAttrName].(string)
+			lset, _ := utils.LabelsFromString(lsetAttr)
+			currentResultHash := lset.Hash()
+
+			// Aggregating cross series aggregates, only supported over raw data.
+			if ctx.isCrossSeriesAggregate {
+				lastTimePerMetric[currentResultHash], lastValuePerMetric[currentResultHash], _ = aggregateClientAggregatesCrossSeries(ctx, res, lastTimePerMetric[currentResultHash], lastValuePerMetric[currentResultHash])
+				// Aggregating over time aggregates
```
Comment (gtopper) [inline pkg/pquerier/collector.go]: What's this comment doing here at the end of a block?

File: pkg/pquerier/collector.go [hunk 2]
```diff
@@ -204,3 +215,57 @@ func downsampleRawData(ctx *selectQueryContext, res *qryResults,
 
 	return lastT, lastV, nil
 }
+
+func aggregateClientAggregatesCrossSeries(ctx *selectQueryContext, res *qryResults, previousPartitionLastTime int64, previousPartitionLastValue float64) (int64, float64, error) {
+	ctx.logger.Debug("using Client Aggregates Collector for metric %v", res.name)
+	it := newRawChunkIterator(res, nil).(*rawChunkIterator)
```
Comment (gtopper) [inline pkg/pquerier/collector.go]: Pass the logger in? `newRawChunkIterator` should already return a `rawChunkIterator`.

File: pkg/pquerier/collector.go (hunk 1 above)
Comment (talIguaz) [reply pkg/pquerier/collector.go]: it's supposed to be for the `else`, but for some reason this is how go formatted it

File: pkg/pquerier/collector.go [hunk 3]
```diff
@@ -204,3 +215,57 @@ func downsampleRawData(ctx *selectQueryContext, res *qryResults,
...
+
+	for currBucket := previousPartitionEndBucket; currBucket < maxBucketForPartition; currBucket++ {
+		currBucketTime := int64(currBucket)*ctx.queryParams.Step + ctx.queryParams.From
+
+		if it.Seek(currBucketTime) {
+			t, v := it.At()
+			if t == currBucketTime {
+				for _, col := range res.frame.columns {
+					if col.GetColumnSpec().metric == res.name {
+						col.SetDataAt(currBucket, v)
```
Comment (gtopper) [inline pkg/pquerier/collector.go]: If we're going to ignore errors, we should use `_ = ` notation to explicitly do so (I'm surprised the linter is ok with this - I think that's not the case in Nuclio).

File: pkg/pquerier/collector.go (hunk 1 above)
Comment (gtopper) [reply pkg/pquerier/collector.go]: So move it inside the else block?

File: pkg/pquerier/query_integration_test.go [hunk 4]
```diff
@@ -1888,6 +1888,430 @@ func (suite *testQuerySuite) toMillis(date string) int64 {
...
+	adapter, err := tsdb.NewV3ioAdapter(suite.v3ioConfig, nil, nil)
+	suite.Require().NoError(err, "failed to create v3io adapter")
+
+	labels1 := utils.LabelsFromStringList("os", "linux")
+	labels2 := utils.LabelsFromStringList("os", "mac")
+	numberOfEvents := 10
+	eventsInterval := 60 * 1000
+	baseTime := tsdbtest.NanosToMillis(time.Now().UnixNano()) - int64(numberOfEvents*eventsInterval)
+
+	fmt.Println("the base time is: ", baseTime)
```
Comment (gtopper) [inline pkg/pquerier/query_integration_test.go]: ⚡️

File: pkg/pquerier/query_integration_test.go [hunk 5]
```diff
@@ -1888,6 +1888,430 @@ func (suite *testQuerySuite) toMillis(date string) int64 {
...
```
Comment (gtopper) [inline pkg/pquerier/query_integration_test.go]: ⚡️

File: pkg/pquerier/query_integration_test.go [hunk 6]
```diff
@@ -1888,6 +1888,430 @@ func (suite *testQuerySuite) toMillis(date string) int64 {
...
```
Comment (gtopper) [inline pkg/pquerier/query_integration_test.go]: ⚡️

File: pkg/pquerier/collector.go [hunk 7]
```diff
@@ -204,3 +215,57 @@ func downsampleRawData(ctx *selectQueryContext, res *qryResults,
...
+				// In case it's the first point in the partition use the last point of the previous partition for the interpolation
+				if prevT == 0 {
+					prevT = previousPartitionLastTime
+					prevV = previousPartitionLastValue
+				}
+
+				for _, col := range res.frame.columns {
+					if col.GetColumnSpec().metric == res.name {
+						interpolatedT, interpolatedV := col.GetInterpolationFunction()(prevT, t, currBucketTime, prevV, v)
+						if interpolatedT != 0 && interpolatedV != 0 {
```
Comment (gtopper) [inline pkg/pquerier/collector.go]: Bug 🐛🙈

File: pkg/pquerier/collector.go [hunk 8]
```diff
@@ -110,14 +121,14 @@ func rawCollector(ctx *selectQueryContext, res *qryResults) {
 
 func aggregateClientAggregates(ctx *selectQueryContext, res *qryResults) {
 	ctx.logger.Debug("using Client Aggregates Collector for metric %v", res.name)
-	it := newRawChunkIterator(res, nil)
+	it := newRawChunkIterator(res, ctx.logger.GetChild("rawChunkIterator"))
```
Comment (gtopper) [inline pkg/pquerier/collector.go]: This (`GetChild`) is generally done inside the constructor. See [nuclio](https://github.com/nuclio/nuclio/blob/f2ebdd95e95c4a1cb24fcb14e18ea88cd9093eba/pkg/processor/trigger/kinesis/trigger.go#L42) for an example.

File: pkg/pquerier/collector.go [hunk 9]
```diff
@@ -165,7 +176,7 @@ func downsampleRawData(ctx *selectQueryContext, res *qryResults,
 
 	var lastT int64
 	var lastV float64
```
Comment (gtopper) [inline pkg/pquerier/collector.go]: Would be nice to move these as well.

File: pkg/pquerier/interpolate.go [hunk 10]
```diff
@@ -56,46 +73,55 @@ func StrToInterpolateType(str string) (InterpolationType, error) {
...
-}
-
-// linear estimator (smooth graph)
-func projectLinear(tprev, tnext, tseek int64, vprev, vnext float64) (int64, float64) {
-	if math.IsNaN(vprev) || math.IsNaN(vnext) {
-		return tseek, math.NaN()
+func validateTolerance(tolerance, tseek, tprev, tnext int64) bool {
+	// If previous point is too far behind for interpolation or the next point is too far ahead then return NaN
+	if (tprev != 0 && tseek-tprev > tolerance) || tnext-tseek > tolerance {
+		return false
```
Comment (gtopper) [inline pkg/pquerier/interpolate.go]: The comment seems to be wrong.

**Pull request merged**